File: FormsPinViewSample/FormsPinViewSample.PCL/Models/ModelBase.fs
namespace FormsPinViewSample.PCL.Models

open System
open System.ComponentModel
open Microsoft.FSharp.Quotations.Patterns

type ModelBase() = 
    let propertyChanged = Event<_, _>()

    interface INotifyPropertyChanged with
        [<CLIEvent>]
        member this.PropertyChanged = propertyChanged.Publish 

    member this.NotifyPropertyChanged propertyName = 
        propertyChanged.Trigger(this, PropertyChangedEventArgs propertyName)

    member this.NotifyPropertyChanged propertySelector = 
        match propertySelector with 
        | PropertyGet(Some (Value (instance, _)), property, _) when Object.ReferenceEquals(instance, this) -> 
            this.NotifyPropertyChanged property.Name
        | _ -> invalidOp "Expecting property getter expression only (for example `this.SomeProperty`)."

File: FormsPinViewSampleCSharp/FormsPinViewSample.CSharp.Core/Models/ModelBase.cs
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace FormsPinViewSample.Core.Models
{
    public class ModelBase : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;

        protected void RaisePropertyChanged([CallerMemberName] string key = null)
        {
            var propertyChanged = PropertyChanged;
            if (propertyChanged != null)
            {
                propertyChanged(this, new PropertyChangedEventArgs(key));
            }
        }

        protected void RaisePropertiesChanged(params string[] keys)
        {
            if (keys != null)
            {
                foreach (string key in keys)
                {
                    var propertyChanged = PropertyChanged;
                    if (propertyChanged != null)
                    {
                        propertyChanged(this, new PropertyChangedEventArgs(key));
                    }
                }
            }
        }
    }
}